File: PE/ModuleDefinition.ts
/// <reference path="Version.ts" />
/// <reference path="Machine.ts" />

/// <reference path="PEFormat/PEFile.ts" />

/// <reference path="Cli/ClrImageFlags.ts" />

module Mi.PE {
    export class ModuleDefinition {
        pe: PEFormat.PEFile;

        runtimeVersion: Version;

        imageFlags: Mi.PE.Cli.ClrImageFlags;

        metadataVersion: Version;
        metadataVersionString: string;

        tableStreamVersion: Version;


        // Ushort
        generation: number;

        name: string;

        // The mvid column shall index a unique GUID in the GUID heap (ECMA-335 §24.2.5)
        // that identifies this instance of the module.
        // The mvid can be ignored on read by conforming implementations of the CLI.
        // The mvid should be newly generated for every module,
        // using the algorithm specified in ISO/IEC 11578:1996
        // (Annex A) or another compatible algorithm.

        // [Rationale: While the VES itself makes no use of the Mvid,
        // other tools (such as debuggers, which are outside the scope of this standard)
        // rely on the fact that the <see cref="Mvid"/> almost always differs from one module to another.
        // end rationale]
        mvid: string;

        encId: string;
        encBaseId: string;

        types: TypeDefinition[];
    }
}

File: Zoom.PE/Mi.PE/Cli/ModuleDefinition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Mi.PE.Cli
{
    using Mi.PE.Cli.Tables;
    using Mi.PE.Internal;
    using Mi.PE.PEFormat;

    public sealed class ModuleDefinition
    {
        public Version RuntimeVersion;
        public ClrImageFlags ImageFlags;
        public Version MetadataVersion;
        public string MetadataVersionString;
        public Version TableStreamVersion;

        public TypeDefinition[] Types;
        
        public string Name;
        public ushort Generation;
        public Guid? Mvid;
        public Guid? EncId;
        public Guid? EncBaseId;
    }
}